working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,431 @@
+the scope where the macro is applied, the scope of the generated code (may be
+same as the former), and the scope where the macro is defined. How does it
+distinguish between these?
+
+### Ordering and staging
+
+In what order are macros run? If they can both generate new APIs and introspect
+on the program, then ordering can be user-visible. Consider a macro that adds a
+field for memoization and another that adds a method to serialize all fields. If
+both are applied to the same class, does the memoized field get serialized?
+
+Do macros run before type checking, after, or interleaved? What is the static
+type context in which a macro runs?
+
+How do macros interact with const canonicalization and evaluation, do they run
+before, after, or interleaved? Can macros read const values *and* generate them?
+
+There are additional considerations if we allow recursive macros, and cycles
+between macro defining libraries.
+
+### Usability
+
+Macros can easily lead to incomprehensible programs if they become overly
+magical, or are a complete black box to the user. In order to demystify macros
+you should:
+
+* Be able to easily navigate to the macro implementation.
+* Be able to visualize in some way the code generated into your program, at
+  development time.
+* Be able to debug generated code (step through it, etc).
+  * Possibly provide an opt out for this, if feasible/desirable.
+* Be able to trace errors that flow through generated code, as well as navigate
+  back to the line in the macro implementation that produced the code.
+* Be able to auto-complete macro generated apis.
+
+### Performance
+
+A Turing-complete programming language that runs in your typechecker opens the
+door to user-code that locks the IDE. How do we ensure that users maintain a
+fast edit refresh cycle when arbitrary Dart code may be running during compilation?
+
+[function_builders]: https://github.com/apple/swift-evolution/blob/9992cf3c11c2d5e0ea20bee98657d93902d5b174/proposals/XXXX-function-builders.md
+[zig_compile_time_execution]: https://andrewkelley.me/post/zig-programming-language-blurs-line-compile-time-run-time.html#:~:text=Compile%2DTime%20Parameters,-Compile%2Dtime%20parameters&text=In%20Zig%2C%20types%20are%20first,functions%2C%20and%20returned%20from%20functions.&text=At%20the%20callsite%2C%20the%20value,is%20known%20at%20compile%2Dtime.
+[data_classes_issue]: https://github.com/dart-lang/language/issues/314
+[kotlin_copy]: https://kotlinlang.org/docs/reference/data-classes.html#copying
+[observable]: https://docs.google.com/document/d/1L2NgM-Kl1PKt8iWQIrFugH0JUQ7Ky86MrM-WM1CPf0I/edit#heading=h.sb9uvy8myyqv
+[differentiable]: https://en.wikipedia.org/wiki/Differentiable_programming
+[rust_declarative_macros]: https://doc.rust-lang.org/book/ch19-06-macros.html#declarative-macros-with-macro_rules-for-general-metaprogramming
+[rust_procedural_macros]: https://doc.rust-lang.org/book/ch19-06-macros.html#procedural-macros-for-generating-code-from-attributes
```

Comment:
might want to reference flutter_hooks and react native hooks + possible js features.
https://github.com/flutter/flutter/issues/51752
https://github.com/rrousselGit/flutter_hooks
Would be interesting to understand if the features this doc is discussing are powerful for the use cases @rrousselGit is thinking about.